### PR TITLE
feat(providers): add WeChat

### PR DESCRIPTION
**Pull request by zhuangya**
## Reasoning 💡
introduce WeChat OAuth 2.0 provider

## Checklist 🧢

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

File: src/server/lib/oauth/client.js [hunk 1]
```diff
@@ -131,6 +131,9 @@ async function getOAuth2AccessToken (code, provider, codeVerifier) {
     { algorithm: 'ES256', keyid: keyId }
     )
     params.client_secret = clientSecret
+  } else if (provider.id === 'wechat') {
+    params.appid = provider.clientId || provider.appid
+    params.secret = provider.clientSecret;
```
Comment (balazsorban44) [inline src/server/lib/oauth/client.js:136]: Custom behavior like this won't be supported in the upcoming major release. Read #2411 and the corresponding RFC #1846 for context.

File: src/server/lib/oauth/client.js [hunk 2]
```diff
@@ -232,6 +235,17 @@ async function getOAuth2 (provider, accessToken, results) {
...
+
+    // WeChat has different url params design
+    if (provider.id === 'wechat') {
+      const wechatProfileURL = new URL(url)
+
+      wechatProfileURL.searchParams.append('access_token', results.accessToken)
+      wechatProfileURL.searchParams.append('openid', results.openid)
+
+      url = wechatProfileURL.href
+    }
```
Comment (balazsorban44) [inline src/server/lib/oauth/client.js:247]: Custom behavior like this won't be supported in the upcoming major release. Read #2411 and the corresponding RFC #1846 for context.

File: src/providers/wechat.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    accessTokenUrl: "https://api.weixin.qq.com/sns/oauth2/access_token",
+    authorizationUrl:
+      "https://open.weixin.qq.com/connect/oauth2/authorize?response_type=code",
+    profileUrl: "https://api.weixin.qq.com/sns/userinfo",
+
+    async profile(profile) {
+      return {
+        id: profile.openid,
+        nickname: profile.nickname,
+        avatar: profile.headimgurl
```
Comment (balazsorban44) [inline src/providers/wechat.js:24]: The upcoming version will require all built-in providers to return a unified output here by default, which is `id`, `name`, `email`, `image`, defaulting to `null`. Read #2411 and the corresponding RFC #1846 for context.

File: www/docs/providers/wechat.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+---
+id: wechat
+title: WeChat
+---
+
+## Documentation
+
+https://developers.weixin.qq.com/doc/offiaccount/en/OA_Web_Apps/Wechat_webpage_authorization.html
+
```
Comment (balazsorban44) [inline www/docs/providers/wechat.md:9]: Please also add a Configuration section

File: src/providers/wechat.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    name: "WeChat",
+    type: "oauth",
+    version: "2.0",
+    scope: options.scope,
+    params: { grant_type: "authorization_code" },
+    authorizationParams: {
+      appid: options.clientId
+    },
+    protection: "state",
```
Comment (balazsorban44) [inline src/providers/wechat.js:13]:
```suggestion
    protection: ["state"],
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thank you for this PR! We are close to a new major release, so I kindly ask you to change the base of this PR from `main` to `next`. Unfortunately, there is a big chance for a few merge conflicts, so the easiest way is to probably to reopen this PR where you base your branch on our `next` branch to begin with. I am sorry for the inconvenience, this will hopefully not be required for long.

To comply with the new Provider API, I kindly ask you to incorporate my suggestions. I might have a few more when you have a PR against `next`, but I am happy to help you through this to be merged!

I would also like to ask, do you know how we could set up a test account on WeChat to add it to our basic sign-in test suite to ship this with more confidence?

Feel free to ask me anything

**Comment by zhuangya** [comment]
> Thank you for this PR! We are close to a new major release, so I kindly ask you to change the base of this PR from `main` to `next`. Unfortunately, there is a big chance for a few merge conflicts, so the easiest way is to probably to reopen this PR where you base your branch on our `next` branch to begin with. I am sorry for the inconvenience, this will hopefully not be required for long.
> 
> To comply with the new Provider API, I kindly ask you to incorporate my suggestions. I might have a few more when you have a PR against `next`, but I am happy to help you through this to be merged!
> 

No worry, I'll take care of this.
I've already gone through your suggestions, and I'd like to read #2411 and #1846 first.

> I would also like to ask, do you know how we could set up a test account on WeChat to add it to our basic sign-in test suite to ship this with more confidence?

not yet, but i think i can figure it out

> Feel free to ask me anything

Thank you, it's very kind of you!

**Comment by jonidelv** [comment]
I really need this 😿 is there any chance of this moving forward? Sorry, I'm not yet familiar with the code to help.

**Comment by StarDxxx** [comment]
we need wechat

**Comment by DaZiYuan** [comment]
Please support WeChat, WeChat is a commonly used way in China

**Comment by kulame** [comment]
we need wechat., please support it !

**Comment by zuohuadong** [comment]
we need wechat., please support it !

**Comment by sharkqwy** [comment]
World's 18% users are on WeChat. It doesn't make sense to support all other smaller social providers while ignore it. Please add support for WeChatProvider

**Comment by daniel548604106** [comment]
Please support wechat

**Comment by JacobSyndeo** [comment]
WeChat is even more important to support than one may think, as many of the other top providers (e.g. Google, Facebook, X/Twitter, Discord, etc.) are _completely blocked_ in mainland China, leaving WeChat as essentially _THE_ only viable option there.

It's not just a matter of WeChat being installed on nearly 1 billion phones… it's that it's the de-facto platform in China, which is (in a close tie for) the most populated nation in the world.

So yeah, WeChat matters. Please prioritize this.

(If it's a matter of not being able to test it, I can help test this if necessary, and I'm sure at least _some_ of the rest of the 1 billion users will help test it as well!)

**Pull request closed without merging**